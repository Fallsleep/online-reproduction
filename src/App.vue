<script setup lang="ts">
import { App as AApp } from 'ant-design-vue';
</script>

<template>
  <a-config-provider>
    <a-app>
      <app-containter>
        <router-view />
      </app-containter>
    </a-app>
  </a-config-provider>
</template>

<style scoped></style>
