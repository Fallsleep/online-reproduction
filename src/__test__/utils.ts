import { mount } from '@vue/test-utils';

let globalWrapper: ReturnType<typeof mount>;

/**
 * 临时保存 mount 返回的对象
 * @param wrapper mount 返回的对象
 */
export const setWrapper = (wrapper: ReturnType<typeof mount>) => {
  globalWrapper = wrapper;
};

/**
 * 获取 mount/testMount 返回的对象
 */
export const getWrapper = () => globalWrapper;

/**
 * 调用 mount/testMount 返回的对象的 unmount 方法
 */
export const cleanup = () => {
  if (globalWrapper) {
    globalWrapper.unmount();
    globalWrapper = null;
  }
};
