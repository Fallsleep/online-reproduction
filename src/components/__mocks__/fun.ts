export const fun = vi.fn();
