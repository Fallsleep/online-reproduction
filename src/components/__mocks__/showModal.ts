export const showModal = vi.fn();
