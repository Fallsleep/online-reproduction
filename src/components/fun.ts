export const fun = (str: string) => {
  return new Promise(() => {
    return `param: ${str}`;
  });
};
