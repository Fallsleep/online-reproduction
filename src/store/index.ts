import type { PiniaPlugin } from 'pinia';
import type { Router } from 'vue-router';

export const pinia = createPinia();

export const pinaRouterPlugin = (router: Router): PiniaPlugin => {
  return ({ store }) => {
    store.router = markRaw(router);
  };
};
