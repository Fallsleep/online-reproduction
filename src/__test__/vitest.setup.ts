import { server } from '@/mock/server';
import {
  VueRouterMock,
  createRouterMock,
  injectRouterMock,
} from 'vue-router-mock';
import { config } from '@vue/test-utils';
import { cleanup } from './utils';

// Establish API mocking before all tests.
beforeAll(() => {
  server.listen();
});

const router = createRouterMock();
beforeEach(() => {
  injectRouterMock(router);
});
// Reset any request handlers that we may add during the tests,
// so they don't affect other tests.
afterEach(() => {
  server.resetHandlers();
  vi.restoreAllMocks();
  cleanup();
});
// Clean up after the tests are finished.
afterAll(() => {
  server.close();
});
config.plugins.VueWrapper.install(VueRouterMock);
