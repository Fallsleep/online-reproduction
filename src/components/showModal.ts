import { useAntdvStore } from '@/store/antdv';
import { fun } from './fun';

export const showModal = (anyString: string) => {
  useAntdvStore().modal.confirm({
    title: '提示',
    content: '你确定要操作么？',
    okText: '确定',
    cancelText: '取消',
    onOk: () => {
      console.log(111);
      return fun(anyString).then((res) => console.log(res));
    },
  });
};
