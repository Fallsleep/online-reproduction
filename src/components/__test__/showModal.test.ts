import { fun } from '../fun';
import { showModal } from '../showModal';
import { pinaRouterPlugin } from '@/store';
import App from '@/App.vue';
import { setWrapper } from '@/__test__/utils';

describe('对话框组件测试', () => {
  vi.mock('../fun');
  beforeEach(() => {
    setWrapper(
      testMount(App, {
        global: {
          plugins: [
            // 和 @testing-library/vue 不同，不需要将 router 作为插件传递
            createTestingPinia({
              // 由于 vue-router-mock 是在 vitest.setup.js的 beforeEach 中注入的 router 对象，
              // 必须在 beforeEach 或 it/test 中调用 getRouter 获取router，否则是 undefined。
              plugins: [pinaRouterPlugin(getRouter())],
            }),
          ],
        },
        attachTo: document.body,
      })
    );
  });

  it('显示对话框，快照测试', async () => {
    // 弹出对话框，除非是在 beforeEach 中调用，此处必须用 await ，否则后续将无法获取到组件
    await showModal();
    // 是否显示了对话框
    const message = testScreen.getByText('你确定要操作么？');
    // 显示对话框
    expect(message).toBeTruthy();
    // 快照
    expect(document.body).toMatchSnapshot();
  });

  it('点击确定按钮，调用 fun 方法', async () => {
    // 弹出对话框，除非是在 beforeEach 中调用，此处必须用 await ，否则后续将无法获取到组件
    await showModal('any string');
    const user = userEvent.setup();
    // 点击确定按钮
    // 1. 按钮中文见会加空格
    // 2. 目前文字父节点为 span ，而非注册了点击事件的 button
    await user.click(
      testScreen.getByRole('button', {
        name: /确 定/i,
        hidden: true,
      })
    );
    // 调用了 fun 方法
    expect(fun).toHaveBeenCalledOnce();
    expect(fun).toHaveBeenCalledWith('any string');
  });
});
