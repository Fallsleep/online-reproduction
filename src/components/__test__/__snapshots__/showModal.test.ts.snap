// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`对话框组件测试 > 显示对话框，快照测试 1`] = `
<body>
  <div
    data-v-app=""
  >
    
    
    
    <div
      class="css-dev-only-do-not-override-kqecok ant-app"
    >
      
      <!--teleport start-->
      <!--teleport end-->
      
      <!--teleport start-->
      <!--teleport end-->
      <!--teleport start-->
      <!--teleport end-->
      
      <div>
        
        <router-view-stub
          name="default"
        />
        
      </div>
      
    </div>
    
    
    
  </div>
  
  
  
  
  
  
  <div>
    <div
      class="ant-modal-root css-dev-only-do-not-override-kqecok"
    >
      <transition-stub
        appear="true"
        css="true"
        enteractiveclass="ant-fade-enter ant-fade-enter-prepare"
        enterfromclass="ant-fade-enter ant-fade-enter-prepare ant-fade-enter-start"
        entertoclass="ant-fade-enter ant-fade-enter-active"
        leaveactiveclass="ant-fade-leave ant-fade-leave-active"
        leavefromclass=" ant-fade-leave"
        leavetoclass="ant-fade-leave ant-fade-leave-active"
        name="ant-fade"
        persisted="false"
      >
        <div
          class="ant-modal-mask"
        />
      </transition-stub>
      <div
        class="ant-modal-wrap"
        role="dialog"
      >
        <transition-stub
          appear="true"
          css="true"
          enteractiveclass="ant-zoom-enter ant-zoom-enter-prepare"
          enterfromclass="ant-zoom-enter ant-zoom-enter-prepare ant-zoom-enter-start"
          entertoclass="ant-zoom-enter ant-zoom-enter-active"
          leaveactiveclass="ant-zoom-leave ant-zoom-leave-active"
          leavefromclass=" ant-zoom-leave"
          leavetoclass="ant-zoom-leave ant-zoom-leave-active"
          name="ant-zoom"
          persisted="false"
        >
          <div
            class="ant-modal css-dev-only-do-not-override-kqecok ant-modal-confirm ant-modal-confirm-confirm"
            role="document"
            style="width: 416px;"
          >
            <div
              aria-hidden="true"
              style="width: 0px; height: 0px; overflow: hidden; outline-color: none; outline-style: none; outline-width: initial;"
              tabindex="0"
            />
            <div
              class="ant-modal-content"
            >
              <!---->
              <!---->
              <div
                class="ant-modal-body"
              >
                
                <div
                  class="ant-modal-confirm-body-wrapper"
                >
                  <div
                    class="ant-modal-confirm-body"
                  >
                    <span
                      aria-label="exclamation-circle"
                      class="anticon anticon-exclamation-circle"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="exclamation-circle"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm-32 232c0-4.4 3.6-8 8-8h48c4.4 0 8 3.6 8 8v272c0 4.4-3.6 8-8 8h-48c-4.4 0-8-3.6-8-8V296zm32 440a48.01 48.01 0 010-96 48.01 48.01 0 010 96z"
                        />
                      </svg>
                      <!---->
                    </span>
                    <span
                      class="ant-modal-confirm-title"
                    >
                      提示
                    </span>
                    <div
                      class="ant-modal-confirm-content"
                    >
                      你确定要操作么？
                    </div>
                  </div>
                  <div
                    class="ant-modal-confirm-btns"
                  >
                    <button
                      class="css-dev-only-do-not-override-kqecok ant-btn ant-btn-default"
                      type="button"
                    >
                      <transition-stub
                        appear="false"
                        css="true"
                        name="ant-btn-loading-icon-motion"
                        persisted="false"
                      >
                        <!---->
                      </transition-stub>
                      
                      <span>
                        取 消
                      </span>
                      
                    </button>
                    <button
                      class="css-dev-only-do-not-override-kqecok ant-btn ant-btn-primary"
                      type="button"
                    >
                      <transition-stub
                        appear="false"
                        css="true"
                        name="ant-btn-loading-icon-motion"
                        persisted="false"
                      >
                        <!---->
                      </transition-stub>
                      
                      <span>
                        确 定
                      </span>
                      
                    </button>
                  </div>
                </div>
                
              </div>
              <!---->
            </div>
            <div
              aria-hidden="true"
              style="width: 0px; height: 0px; overflow: hidden; outline-color: none; outline-style: none; outline-width: initial;"
              tabindex="0"
            />
          </div>
        </transition-stub>
      </div>
    </div>
    
  </div>
</body>
`;
